File: app/schemas/ai.py
from pydantic import BaseModel, Field
from typing import Optional, List
from app.schemas.intent import IntentContractCreate

class ChatMessage(BaseModel):
    role: str = Field(..., description="'user' or 'assistant'")
    content: str = Field(..., description="Message text")

class AIIntentRequest(BaseModel):
    text: str = Field(..., description="The natural language user request to interpret.")
    history: Optional[list[ChatMessage]] = Field(default_factory=list, description="Previous conversation history")

class AIIntentResponse(BaseModel):
    intent: Optional[IntentContractCreate] = Field(None, description="The strictly structured intent extracted from the request.")
    interpretation: str = Field(..., description="Human-readable explanation of what the AI interpreted.")
    success: bool = Field(..., description="Whether extraction and strict validation was successful.")
    error: Optional[str] = Field(None, description="Error message if validation failed.")

class ChatRequest(BaseModel):
    message: str = Field(..., description="The user's chat message.")
    history: Optional[List[ChatMessage]] = Field(default_factory=list, description="Conversation history")

class ChatResponse(BaseModel):
    reply: str = Field(..., description="The AI's response.")
    success: bool = True
    error: Optional[str] = None

File: frontend/src/services/api/ai.ts
import { apiClient } from '../api';

export interface MerchantRestrictions {
  type: "NONE" | "ALLOWLIST" | "BLOCKLIST";
  list: string[];
}

export interface ExtractedIntent {
  currency: string;
  max_total_amount: number;
  allowed_categories: string[];
  banned_categories: string[];
  max_quantity?: number | null;
  recurring_payment_allowed: boolean;
  merchant_restrictions: MerchantRestrictions;
}

export interface IntentExtractionResponse {
  intent: ExtractedIntent | null;
  interpretation: string;
  success: boolean;
  error?: string;
}

export interface ChatMessage {
  role: 'user' | 'assistant';
  content: string;
}

export interface ChatResponse {
  reply: string;
  success: boolean;
  error?: string;
}

export const aiService = {
  /**
   * Extract intent from natural language or chat
   */
  extractIntent: async (text: string, history: ChatMessage[] = []): Promise<IntentExtractionResponse> => {
    const response = await apiClient.post('/ai/extract', { text, history });
    return response.data;
  },

  /**
   * Finance-expert chat with live account context
   */
  chat: async (message: string, history: ChatMessage[] = []): Promise<ChatResponse> => {
    const response = await apiClient.post('/ai/chat', { message, history });
    return response.data;
  }
};
